sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java

```diff
@@ -0,0 +1,488 @@
+      return new UnboundedJmsReader(this, checkpointMark);
+    }
+
+    @Override
+    public void validate() {
+      if (connectionFactory == null) {
+        throw new IllegalArgumentException("ConnectionFactory is not defined");
+      }
+      if (queue == null && topic == null) {
+        throw new IllegalArgumentException("Either queue or topic is not defined");
+      }
+    }
+
+    @Override
+    public Coder getCheckpointMarkCoder() {
+      return SerializableCoder.of(JmsCheckpointMark.class);
+    }
+
+    @Override
+    public Coder<JmsRecord> getDefaultOutputCoder() {
+      return SerializableCoder.of(JmsRecord.class);
+    }
+
+  }
+
+  private static class UnboundedJmsReader extends UnboundedReader<JmsRecord> {
+
+    private UnboundedJmsSource source;
+    private JmsCheckpointMark checkpointMark;
+    private String destination;
+    private Connection connection;
+    private Session session;
+    private MessageConsumer consumer;
+
+    private JmsRecord currentRecord;
+    private Instant currentTimestamp;
+
+    public UnboundedJmsReader(
+        UnboundedJmsSource source,
+        @Nullable JmsCheckpointMark checkpointMark) {
+      this.source = source;
+      this.checkpointMark = checkpointMark;
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      ConnectionFactory connectionFactory = source.connectionFactory;
+      try {
+        this.connection = connectionFactory.createConnection();
+        this.connection.start();
+        this.session = this.connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
+        if (source.topic != null) {
+          this.consumer = this.session.createConsumer(this.session.createTopic(source.topic));
+          this.destination = source.topic;
+        } else {
+          this.consumer = this.session.createConsumer(this.session.createQueue(source.queue));
+          this.destination = source.queue;
+        }
+
+        return advance();
+      } catch (Exception e) {
+        throw new IOException(e);
+      }
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      try {
+        TextMessage message = (TextMessage) this.consumer.receiveNoWait();
+
+        if (message == null) {
+          return false;
+        }
+
+        Map<String, Object> headers = new HashMap<>();
+        headers.put("JMSMessageID", message.getJMSMessageID());
+        headers.put("JMSTimestamp", message.getJMSTimestamp());
+        headers.put("JMSCorrelationID", message.getJMSCorrelationID());
+        headers.put("JMSReplyTo", message.getJMSReplyTo());
+        headers.put("JMSDestination", message.getJMSDestination());
+        headers.put("JMSDeliveryMode", message.getJMSDeliveryMode());
+        headers.put("JMSRedelivered", message.getJMSRedelivered());
+        headers.put("JMSType", message.getJMSType());
+        headers.put("JMSExpiration", message.getJMSExpiration());
+        headers.put("JMSPriority", message.getJMSPriority());
```

Comment:
Is it worth either:

1. extracting these well-known values into their own fields, instead of putting them in headers? [or having a JsmHeaders struct]?
2. Making these strings constants, such as `JMS_HEADER_MESSAGE_ID` so that it is easy for users to re-use them without fear of typos?